### PR TITLE
let vs let!

### PR DESCRIPTION
**Pull request by sebastiancaraballo**
[Discussion Link](https://github.com/rootstrap/tech-guides/issues/75)

**Description:**
- Add `ruby/rspec/tutorial/let-vs-let.md`
- Add reference at `ruby/rspec/tutorial.md`

### CODE REVIEW

File: ruby/rspec/tutorial/let-vs-let.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# let vs let!
+
+Understanding the difference between `let` and `let!` is fundamental when starting to learn RSpec.
+
+Use let to define a memoized helper method. The value will be cached
+across multiple calls in the same example but not across examples.
+
+Note that let is lazy-evaluated: it is not evaluated until the first time
+the method it defines is invoked. You can use let! to force the method's
```
Comment (juanmferreira93) [inline ruby/rspec/tutorial/let-vs-let.md:9]:
```suggestion
the method who defines is invoked. You can use let! to force the method's
```

File: ruby/rspec/tutorial/let-vs-let.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+- When there's a need to trigger an insertion into the database before running an example block.
+- An alternative to `let!` is invoking the variable from a `before` or `subject` block, or even from inside an example. This practice is discouraged.
+
+### Example
+
+#### Bad
+```
+let!(:user) { create(:user) }
+
+let!(:membership) { create(:membership, user: user) }
```
Comment (juanmferreira93) [inline ruby/rspec/tutorial/let-vs-let.md:22]: maybe you could add an example of a non active record creation, like `let!(:params) { { 'some_params': 'first_param' } }` or something like that, to show other cases of wrong and good examples

**Review by juanmferreira93 — APPROVED** [review]
🚀 good job

File: ruby/rspec/tutorial/let-vs-let.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# let vs let!
+
+Understanding the difference between `let` and `let!` is fundamental when starting to learn RSpec.
+
+Use let to define a memoized helper method. The value will be cached
```
Comment (megatux) [inline ruby/rspec/tutorial/let-vs-let.md:5]: in this paragraph "let" should be quoted as above : "Use `let` ..."

File: ruby/rspec/tutorial/let-vs-let.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+across multiple calls in the same example but not across examples.
+
+Note that let is lazy-evaluated: it is not evaluated until the first time
```
Comment (megatux) [inline ruby/rspec/tutorial/let-vs-let.md:8]: Quote _let_ : "Note that `let` ..."

File: ruby/rspec/tutorial/let-vs-let.md (hunk 1 above)
Comment (megatux) [inline ruby/rspec/tutorial/let-vs-let.md:9]: "the method it defines" mm maybe we could change "method" to _expression_?

Comment (megatux) [reply ruby/rspec/tutorial/let-vs-let.md:9]: Quote `let!` here, too

File: ruby/rspec/tutorial/let-vs-let.md (hunk 2 above)
Comment (megatux) [reply ruby/rspec/tutorial/let-vs-let.md:22]: yeah, let's try not to associate `let!` with ActiveRecord usages only

File: ruby/rspec/tutorial/let-vs-let.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# let vs let!
```
Comment (santib) [inline ruby/rspec/tutorial/let-vs-let.md:1]: maybe we can mention that by default use `let`, and only use `let!` if you need it to be executed right-away ?

**Review by santib — APPROVED** [review]
👏

**Pull request merged**